docs/design/deviceshifu/design-deviceShifu-rtsp.md

```diff
@@ -0,0 +1,122 @@
+
+## Goal
+
+Create a deviceShifu-RTSP using third-party library and allow user to connect RTSP server, and record the rtsp stream into file like mp4 / flv and so on, user can export file out of k8s.
+
+## General Design
+
+deviceShifu-RTSP will receive RESTful style request like other, then transfer the request into RTSP method calls and send it to RTSP Server, meanwhile deviceShifu-RTSP provide a RTSP address which forward the RTSP stream from the target RTSP server.
+
+## Detailed Design
+
+### Protocol Specification
+
+deviceShifu-RTSP provide a RTSP server to forward RTSP stream and HTTP Server to handle Request.
+
+```go
+type ProtocolSettings struct {
+    MQTTSetting *MQTTSetting `json:"MQTTSetting,omitempty"`
+    ....
++   RTSPSetting *RTSPSetting `json:"RTSPSetting,omitempty"`
+}
+
+type RTSPSetting struct {
+    Username        string `json:"username,omitempty"`
+    PasswordSecret  string `json:"passwordSecret,omitempty"`
+}
+```
+
+If user turn on Record, deviceShifu will using telemetry Service to push the rtsp stream to telemetry Service, and telemetry Service will forward the stream to the rtsp recording server to record the rtsp video into file.
+```yaml
+...
+telemetry:
+  telemetry:
+    push_stream:
+      instruction: stream
+      pushSettings:
+        telemetryCollectionService: rtsp_record_telemetry
+        kind: stream
+...
+```
+Create PVC and PV need Separate with other deploy to avoid user delete pv by mistake. 
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: <devicename>-pv
+  namespace: deviceshifu
+  labels:
+    pv: <devicename>-pv
+spec:
+  capacity:
+    storage: 10Gi
+  volumeMode: Filesystem
+  accessModes:
+    - ReadWriteOnce
+  persistentVolumeReclaimPolicy: Retain
+  storageClassName: slow
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: <devicename>-pvc
+spec:
+  resources:
+    requests:
+      storage: 10Gi
+  volumeMode: Filesystem
+  accessModes:
+    - ReadWriteOnce
+  selector:
+    matchLabels:
+      pv: <devicename>-pv
+```
+recording server deployment.yaml
+```yaml
+# deployment.yaml
+...
+volumeMounts:
+- name: storage-pv
+  mountPath: /data
+...
+volumns:
+- name: storge-pv
+  persistentVolumeClaim:
+    claimName:  pvc-name
+    optional: true
+```
+### Recording Logic
+
+```mermaid
+flowchart LR
+  dv[device]
+  subgraph k8s
+    ds[deviceshifu-RTSP]
+    ts[telemetryService]
+    rrs[RTSP Recording Server]
+    pv[Persistent Volumes]
+  end
+  dv --> ds --> ts --> rrs --> pv
+```
+
+We need to store the video in pieces, each clip is 1 hour by default, the user can set this option
+
+ffmpeg example:
+```bash
+ffmpeg -i rtsp://[RTSP_URL] -c copy -map 0 -segment_time 300 -f segment video/output%03d.mp4
+```
+### Serving requests
+
+deviceShifu-RTSP would take RESTful-style requests just as other deviceShifu do.
+RTSP support like `stream` / `describe` / `record/start` / `record/stop` requests.
```

Comment:
If we are serving streaming data via telemetry service, would the latency be too high?

Reply:
I think its ok, for it is only for forwarding the stream and not parsing